modules/core/arrow-validation/src/main/kotlin/arrow/validation/refinedTypes/numeric/NonZero.kt

```diff
@@ -0,0 +1,51 @@
+package arrow.validation.refinedTypes.numeric
+
+import arrow.Kind
+import arrow.core.Either
+import arrow.core.EitherPartialOf
+import arrow.data.Nel
+import arrow.data.Validated
+import arrow.data.ValidatedPartialOf
+import arrow.extension
+import arrow.instances.either.applicativeError.applicativeError
+import arrow.instances.nonemptylist.semigroup.semigroup
+import arrow.instances.validated.applicativeError.applicativeError
+import arrow.typeclasses.ApplicativeError
+import arrow.validation.RefinedPredicateException
+import arrow.validation.Refinement
+
+internal fun <A : Number> isNonZero(a: A): Boolean = a != 0
+
+interface NonZero<F, A : Number> : Refinement<F, A> {
+
+  override fun A.refinement(): Boolean = isNonZero(this)
```

Comment:
All of these are lawful type classes. We should consider including the laws of each one of them or just for `Refinement` for now.
That will give users a framework to test their new refined rules that are composed with multiple refinements. No need to be in this PR but would be great to have.